python/ray/dag/dag_node_operation.py

```diff
@@ -368,3 +505,72 @@ def _generate_actor_to_execution_schedule(
     for _, candidates in actor_to_candidates.items():
         assert len(candidates) == 0
     return actor_to_execution_schedule
+
+
+def _optimize_execution_schedule(
+    actor_to_execution_schedule: Dict[
+        "ray.actor.ActorHandle", List[_DAGOperationGraphNode]
+    ],
+    overlap_gpu_communication: bool,
+) -> Dict["ray.actor.ActorHandle", List[_DAGOperationGraphNode]]:
+    """
+    Optimize the execution schedule by overlapping computation and communication.
+
+    Args:
+        actor_to_execution_schedule: A dictionary that maps an actor handle to
+            the execution schedule which is a list of operations to be executed.
+        overlap_gpu_communication: Whether to overlap GPU communication with
+            computation. If False, the function will return the original execution
+            schedule (i.e., actor_to_execution_schedule).
+    """
+    if not overlap_gpu_communication:
+        return actor_to_execution_schedule
+
+    actor_to_optimized_schedule: Dict[
+        "ray.actor.ActorHandle", List[_DAGOperationGraphNode]
+    ] = copy.deepcopy(actor_to_execution_schedule)
+    for optimized_schedule in actor_to_optimized_schedule.values():
+        for i in range(len(optimized_schedule)):
+            if (
+                optimized_schedule[i].operation.type == _DAGNodeOperationType.READ
+                and optimized_schedule[i].requires_nccl
+            ):
+                # For each NCCL read operation (i.e., recv), scan backwards
+                # to find the nearest compute node to swap with so that
+                # the NCCL read operation can be overlapped with computation.
+                for j in range(i - 1, -1, -1):
+                    if (
+                        optimized_schedule[j].operation.type
+                        == _DAGNodeOperationType.COMPUTE
```

Comment:
Note, you'll probably need to update this to not reorder with collective ops now that #47621 is merged.

Reply:
Thanks. For now I'm throwing an error to indicate overlap_gpu_communication does not work with collectives yet. I will update the algorithm when we support collectives.